### PR TITLE
Add new bb cache strategy

### LINKED ISSUE #59: Compare w/ real BB cache
We should compare current BBcache implementation, based on fully-associate algorithm with one of the real cache algorithm with replacement policy (LRU, LFU...)

### PR DESCRIPTION
**Pull request by Tako-San**
Closes #59

### CODE REVIEW

**Comment by Bryanskiy** [comment]
LGBT

**Review by Bryanskiy — COMMENTED** [review]
LGBT

**Review by Bryanskiy — APPROVED** [review]

File: include/hart/hart.hh [hunk 1]
```diff
@@ -16,22 +17,27 @@ namespace sim {
 
 namespace fs = std::filesystem;
 
-using BBCache = std::unordered_map<Addr, BasicBlock>;
+class IBBCache {
+public:
+  virtual ~IBBCache() {}
```
Comment (derzhavin3016) [inline include/hart/hart.hh:22]:
```suggestion
  virtual ~IBBCache() = default;
```

File: include/hart/hart.hh [hunk 2]
```diff
@@ -16,22 +17,27 @@ namespace sim {
...
+  std::unique_ptr<IBBCache> bbc_{};
 
   Memory &getMem() { return state_.mem; };
   Addr &getPC() { return state_.pc; };
 
   BasicBlock createBB(Addr entry);
 
 public:
-  Hart(const fs::path &executable);
+  Hart(const fs::path &executable, std::int64_t bbCacheSize);
```
Comment (derzhavin3016) [inline include/hart/hart.hh:40]:
```suggestion
  Hart(const fs::path &executable, std::ssize_t bbCacheSize);
```
Why don't you use size type for size here? Just to be sure that all values will be appropriate to initialize size of the underlying container

File: src/hart/hart.cc [hunk 3]
```diff
@@ -8,7 +10,67 @@
...
+
+public:
+  const BasicBlock &
+  lookupUpdate(Addr key, std::function<BasicBlock(Addr)> slowGetData) override {
+    if (auto hit = hash_.find(key); hit != hash_.end())
+      return hit->second;
+
+    auto bb = slowGetData(key);
+    return hash_[key] = std::move(bb);
+  }
```
Comment (derzhavin3016) [inline src/hart/hart.cc:24]: Why don't you use `try_emplace()` here in order to reduce the amount of key searching & insertion
If key is not found, you can move bb to the returned iterator

File: src/hart/hart.cc [hunk 4]
```diff
@@ -8,7 +10,67 @@
...
+  LRUCache(std::size_t size) : size_(size) {}
+
+  bool isFull() const { return (cache_.size() == size_); }
+
+  const BasicBlock &
+  lookupUpdate(Addr key, std::function<BasicBlock(Addr)> slowGetData) override {
+    auto hit = hash_.find(key);
+    if (hit == hash_.end()) {
+      if (isFull()) {
+        hash_.erase(cache_.back().first);
```
Comment (derzhavin3016) [inline src/hart/hart.cc:54]: `erase()` method has an overload which accepts iterator to argument to delete

File: src/hart/hart.cc [hunk 5]
```diff
@@ -8,7 +10,67 @@
...
+  using ListIt = typename std::list<std::pair<Addr, BasicBlock>>::iterator;
+  std::unordered_map<Addr, ListIt> hash_{};
+
+public:
+  LRUCache(std::size_t size) : size_(size) {}
+
+  bool isFull() const { return (cache_.size() == size_); }
+
+  const BasicBlock &
+  lookupUpdate(Addr key, std::function<BasicBlock(Addr)> slowGetData) override {
```
Comment (derzhavin3016) [inline src/hart/hart.cc:50]: Consider using `try_emplace` here

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

**Pull request merged**